solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+{
+	if (_declaration)
+	{
+		// XXX the SourceUnit should be the root scope unless we're looking for simple variable identifier.
+		// XXX if vardecl, use decl's scope
+		auto collector = ReferenceCollector(*_declaration);
+		_sourceUnit.accept(collector);
+		return collector.take();
+	}
+
+	return {};
+}
+
+vector<LanguageServer::Location> LanguageServer::references(DocumentPosition _documentPosition)
+{
+	fprintf(stderr, "find all references: %s:%d:%d\n",
+		_documentPosition.uri.c_str(),
+		_documentPosition.position.line,
+		_documentPosition.position.column
+	);
+
+	if (auto const file = m_vfs.find(_documentPosition.uri); file != nullptr)
+	{
+		if (!m_compilerStack)
+			compile(*file);
+
+		solAssert(m_compilerStack.get() != nullptr, "");
+
+		auto const sourceName = file->uri().substr(7); // strip "file://"
+
+		if (auto const sourceNode = findASTNode(_documentPosition.position, sourceName); sourceNode)
+		{
+			auto output = vector<Location>{};
+			if (auto const sourceIdentifier = dynamic_cast<Identifier const*>(sourceNode); sourceIdentifier != nullptr)
+			{
+				auto const declaration = !sourceIdentifier->annotation().candidateDeclarations.empty()
+					? sourceIdentifier->annotation().candidateDeclarations.front()
+					: sourceIdentifier->annotation().referencedDeclaration;
+
+				auto const sourceName = _documentPosition.uri.substr(7); // strip "file://"
+				frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+				for (auto const& highlight: findAllReferences(declaration, sourceUnit))
+				{
+					auto location = Location{};
+					location.range = highlight.range;
+					location.uri = _documentPosition.uri;
+					output.emplace_back(location);
+				}
+			}
+			else if (auto const varDecl = dynamic_cast<VariableDeclaration const*>(sourceNode); varDecl != nullptr)
+			{
+				fprintf(stderr, "AST node is vardecl\n");
+				auto const sourceName = _documentPosition.uri.substr(7); // strip "file://"
+				frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(sourceName);
+				auto const highlights = findAllReferences(varDecl, sourceUnit);
+				for (auto const& highlight: highlights)
+				{
+					auto location = Location{};
+					location.range = highlight.range;
+					location.uri = _documentPosition.uri;
+					output.emplace_back(location);
+				}
+			}
+			else
+			{
+				fprintf(stderr, "not an identifier\n");
+			}
+			return output;
+		}
+		else
+		{
+			fprintf(stderr, "AST node not found\n");
+			// error(_params.requestId, ErrorCode::InvalidParams, "Symbol not found.");
+			return {};
+		}
+	}
+	else
+	{
+		// reply(_params.requestId, output);
+		// error(_params.requestId, ErrorCode::RequestCancelled, "not implemented yet.");
+		return {};
+	}
+	fflush(stderr);
+}
+
+vector<LanguageServer::DocumentHighlight> LanguageServer::semanticHighlight(DocumentPosition _documentPosition)
```

Comment:
What does this do?

Reply:
When you editor cursor is on a symbol that is also somewhere else, all these occurrences are highlighted depending on their use (read or write access). Your C++ LSP should have the same (clangd has it at least).

The distinction between use read/write is not implemented yet, it's possible however (protocol-wise).

Reply:
ah, I see! Can you add a comment in the code?